File: Medium/2657_Find_the_Prefix_Common_Array_of_Two_Arrays/Solution.py
from typing import List, Optional


class Solution:
    def findThePrefixCommonArray(self, A: List[int], B: List[int]) -> List[int]:
        n = len(A)
        ans = []
        seen = set()
        common_count = 0
        
        for i in range(n):
            if A[i] in seen:
                common_count += 1
            else:
                seen.add(A[i])
                
            if B[i] in seen:
                common_count += 1
            else:
                seen.add(B[i])
                
            ans.append(common_count)
            
        return ans
        


if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    A1 = [1,3,2,4]
    B1 = [3,1,2,4]
    print(solution.findThePrefixCommonArray(A1, B1))

    # test cases 2
    A2 = [2,3,1]
    B2 = [3,1,2]
    print(solution.findThePrefixCommonArray(A2, B2))
    

File: Medium/2657_Find_the_Prefix_Common_Array_of_Two_Arrays/Solution.java


public class Solution {
    public int[] findThePrefixCommonArray(int[] A, int[] B) {
        int n = A.length;
        int[] ans = new int[n];
        int[] frequency = new int[n + 1];
        int commonCount = 0;

        for (int i = 0; i < n; i++) {
            if (++frequency[A[i]] == 2) {
                commonCount++;
            }
            if (++frequency[B[i]] == 2) {
                commonCount++;
            }
            ans[i] = commonCount;
        }

        return ans;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        int[] A1 = {1, 3, 2, 4};
        int[] B1 = {3, 1, 2, 4};
        int[] result1 = solution.findThePrefixCommonArray(A1, B1);
        for (int val : result1) System.out.print(val + " ");
        System.out.println();
        
        // test cases 2
        int[] A2 = {2, 3, 1};
        int[] B2 = {3, 1, 2};
        int[] result2 = solution.findThePrefixCommonArray(A2, B2);
        for (int val : result2) System.out.print(val + " ");
        System.out.println();
        
    }
}